src/radical/entk/appman/wfprocessor.py

```diff
@@ -100,145 +110,173 @@ def _initialize_workflow(self):
-                                                   logger=self._logger)
-
-                                        # task_as_dict = json.dumps(executable_task.to_dict())
-                                        workload.append(executable_task)
-
-                                        if executable_stage not in scheduled_stages:
-                                            scheduled_stages.append(
-                                                executable_stage)
-
+                        # If Pipeline is in the final state or suspended, we
+                        # skip processing it.
+                        if pipe.state in states.FINAL or  \
+                           pipe.completed or \
+                           pipe.state == states.SUSPENDED:
+
+                            continue
+
+                        if pipe.state == states.INITIAL:
+
+                            # Set state of pipeline to SCHEDULING if it is
+                            # in INITIAL.
+                            local_transition(obj=pipe,
+                                             obj_type='Pipeline',
+                                             new_state=states.SCHEDULING,
+                                             profiler=self._prof,
+                                             logger=self._logger,
+                                             reporter=self._report)
+
+                        # Get the next stage of this pipeline to process
+                        exec_stage = pipe.stages[pipe.current_stage - 1]
+
+                        if not exec_stage.uid:
+                            # TODO: Move parent uid, name assignment to
+                            # assign_uid function
+                            exec_stage.parent_pipeline['uid'] = pipe.uid
+                            exec_stage.parent_pipeline['name'] = pipe.name
+                            exec_stage._assign_uid(self._sid)
+
+                        # We need to check both INITIAL and SCHEDULED as
+                        # there maybe a task in the stage that needs to be
+                        # resubmitted.
+                        exec_tasks = list()
+                        if exec_stage.state in [states.INITIAL,
+                                                states.SCHEDULED]:
+
+                            # If its a new stage, update its state
+                            if exec_stage.state == states.INITIAL:
+
+                                local_transition(obj=exec_stage,
+                                                    obj_type='Stage',
+                                                    new_state=states.SCHEDULING,
+                                                    profiler=self._prof,
+                                                    logger=self._logger,
+                                                    reporter=self._report)
+
+                            exec_tasks = exec_stage.tasks
+
```

Comment:
The above structure is:
```
if state in [INITIAL, SCHEDULED]:
  if state == INITIAL:
    transition_to_SCHEDULED
  do something
```

I think this would be clearer as:

```
if state == INITIAL:
  transition_to_SCHEDULED

if state == SCHEDULED:
  do something
```
This assumes that `transition_to_SCHEDULED` is atomic.  Is it?

Reply:
Cool, will do. Yes, the assumption is correct.

Reply:
Well, the following is the equivalent:

```
if state == INITIAL:
    transition_to_SCHEDULED
    do something

if state == SCHEDULED:
    do something
```

Would you prefer that? I personally prefer as it exists now, but happy to go with something that works for you.

Reply:
then I don't understand the transition.  I would expect the state to be `SCHEDULED` after, well,  transition to `SCHEDULED`, and then the second `if` would apply and do the something?

```python
if state == INITIAL:
    transition_to_SCHEDULED
    # --> state is now SCHEDULED, and the 'if' below will apply

if state == SCHEDULED:
    do something
```

Reply:
Yes, in the new code, I have implemented it in that manner. Fixed.